hrw.go

```diff
@@ -1,280 +1,124 @@
-		}
-	case []uint:
-		var key = make([]byte, 16)
-		for i := 0; i < length; i++ {
-			binary.BigEndian.PutUint64(key, uint64(slice[i]))
-			rule = append(rule, Hash(key))
-		}
-	case []int8:
-		for i := 0; i < length; i++ {
-			key := byte(slice[i])
-			rule = append(rule, Hash([]byte{key}))
-		}
-	case []uint8:
-		for i := 0; i < length; i++ {
-			key := slice[i]
-			rule = append(rule, Hash([]byte{key}))
-		}
-	case []int16:
-		var key = make([]byte, 8)
-		for i := 0; i < length; i++ {
-			binary.BigEndian.PutUint16(key, uint16(slice[i]))
-			rule = append(rule, Hash(key))
-		}
-	case []uint16:
-		var key = make([]byte, 8)
-		for i := 0; i < length; i++ {
-			binary.BigEndian.PutUint16(key, slice[i])
-			rule = append(rule, Hash(key))
-		}
-	case []int32:
-		var key = make([]byte, 16)
-		for i := 0; i < length; i++ {
-			binary.BigEndian.PutUint32(key, uint32(slice[i]))
-			rule = append(rule, Hash(key))
-		}
-	case []uint32:
-		var key = make([]byte, 16)
-		for i := 0; i < length; i++ {
-			binary.BigEndian.PutUint32(key, slice[i])
-			rule = append(rule, Hash(key))
-		}
-	case []int64:
-		var key = make([]byte, 32)
-		for i := 0; i < length; i++ {
-			binary.BigEndian.PutUint64(key, uint64(slice[i]))
-			rule = append(rule, Hash(key))
-		}
-	case []uint64:
-		var key = make([]byte, 32)
-		for i := 0; i < length; i++ {
-			binary.BigEndian.PutUint64(key, slice[i])
-			rule = append(rule, Hash(key))
-		}
-	case []string:
-		for i := 0; i < length; i++ {
-			rule = append(rule, Hash([]byte(slice[i])))
-		}
+	oHash := object.Hash()
 
-	default:
-		if _, ok := val.Index(0).Interface().(Hasher); !ok {
-			panic("slice elements must implement hrw.Hasher")
-		}
+	var s sliceToSort[V, W]
+	s.s = vv
+	s.distances = make([]W, len(vv))
 
-		for i := 0; i < length; i++ {
-			h := val.Index(i).Interface().(Hasher)
-			rule = append(rule, h.Hash())
-		}
+	for i := range vv {
+		// the distance is a bad characteristic in our case (we sort in ascending order)
+		// so a bigger weight should lower the distance more
+		s.distances[i] = W(distance(vv[i].Hash(), oHash)) / weights[i]
 	}
-	return rule
-}
 
-// ValidateWeights checks if weights are normalized between 0.0 and 1.0.
-func ValidateWeights(weights []float64) error {
-	for i := range weights {
-		if math.IsNaN(weights[i]) || weights[i] > NormalizedMaxWeight || weights[i] < NormalizedMinWeight {
-			return errors.New("weights are not normalized")
-		}
-	}
-	return nil
+	sort.Stable(&s)
 }
 
-func newSorter(l int, byIndex bool, nodes []uint64, h uint64,
-	swap func(i, j int)) (*sorter, []int, []uint64) {
-	ind := make([]int, l)
-	dist := make([]uint64, l)
-	for i := 0; i < l; i++ {
-		ind[i] = i
-		dist[i] = getDistance(byIndex, i, nodes, h)
-	}
+type distancesValue interface {
+	constraints.Unsigned | constraints.Float
+}
 
-	return &sorter{
-		l: l,
-		swap: func(i, j int) {
-			swap(i, j)
-			ind[i], ind[j] = ind[j], ind[i]
-		},
-	}, ind, dist
+type sliceToSort[V any, W distancesValue] struct {
```

Comment:
BTW, this type (and `struct` allocation and conversion to `sort.Interface`) can be avoided if you're to use `sort.SliceStable()` with an appropriate `less`. But I'm not sure which way is gonna perform better, needs to be tested.

Reply:
never really thought about `reflectlite` package but not really think that it can be better than just slice swap

with this diff:

```diff
diff --git a/hrw.go b/hrw.go
index 75fad56..94bba4e 100644
--- a/hrw.go
+++ b/hrw.go
@@ -41,15 +41,14 @@ func (h hashableWrapper) Hash() uint64 {
 func Sort[V, P Hashable](vv []V, object P) {
        oHash := object.Hash()
 
-       var s sliceToSort[V, uint64]
-       s.s = vv
-       s.distances = make([]uint64, len(vv))
-
+       distances := make([]uint64, len(vv))
        for i := range vv {
-               s.distances[i] = distance(vv[i].Hash(), oHash)
+               distances[i] = distance(vv[i].Hash(), oHash)
        }
 
-       sort.Stable(&s)
+       sort.SliceStable(vv, func(i, j int) bool {
+               return distances[i] < distances[j]
+       })
 }
```

i have this results (the new one is your suggestion):
```
goos: linux
goarch: amd64
pkg: github.com/nspcc-dev/hrw
cpu: Intel(R) Core(TM) i7-10510U CPU @ 1.80GHz
                │   old.txt   │                 new.txt                 │
                │   sec/op    │    sec/op      vs base                  │
Sort_fnv_10-8     133.0n ± 2%     207.1n ± 3%    +55.69% (p=0.000 n=10)
Sort_fnv_100-8    923.1n ± 3%     998.7n ± 3%     +8.19% (p=0.000 n=10)
Sort_fnv_1000-8   8.564µ ± 1%   145.805µ ± 0%  +1602.63% (p=0.000 n=10)
geomean           1.017µ          3.113µ        +206.10%

                │   old.txt    │               new.txt                │
                │     B/op     │     B/op      vs base                │
Sort_fnv_10-8       128.0 ± 0%     168.0 ± 0%  +31.25% (p=0.000 n=10)
Sort_fnv_100-8      944.0 ± 0%     984.0 ± 0%   +4.24% (p=0.000 n=10)
Sort_fnv_1000-8   8.047Ki ± 0%   8.086Ki ± 0%   +0.49% (p=0.000 n=10)
geomean             998.5        1.084Ki       +11.19%

                │  old.txt   │               new.txt               │
                │ allocs/op  │ allocs/op   vs base                 │
Sort_fnv_10-8     2.000 ± 0%   4.000 ± 0%  +100.00% (p=0.000 n=10)
Sort_fnv_100-8    2.000 ± 0%   4.000 ± 0%  +100.00% (p=0.000 n=10)
Sort_fnv_1000-8   2.000 ± 0%   4.000 ± 0%  +100.00% (p=0.000 n=10)
geomean           2.000        4.000       +100.00%

```

Reply:
Impressive. Convincing.